crates/hir-ty/src/infer/closure.rs

```diff
@@ -957,7 +957,14 @@ impl InferenceContext<'_> {
         }
         self.restrict_precision_for_unsafe();
         // closure_kind should be done before adjust_for_move_closure
-        let closure_kind = self.closure_kind();
+        let closure_kind = {
+            let from_capture = self.closure_kind_from_capture();
+            // if predicate.unwrap_or(FnTrait::Fn) < from_capture {
+            //     // Diagnostics here, like compiler does in
+            //     // https://github.com/rust-lang/rust/blob/11f32b73e0dc9287e305b5b9980d24aecdc8c17f/compiler/rustc_hir_typeck/src/upvar.rs#L264
+            // }
+            predicate.unwrap_or(from_capture)
```

Comment:
Should use `unwrap_or_else(|| self.closure_kind_from_capture())` for lazyiness

Reply:
I was planning to implement a diagnostic here as the comments above says. I'll try it and will convert into `unwrap_or_else` if it is not possible